### PR TITLE
Fix test flakiness

### PR DESCRIPTION
**Pull request by spacebear21**
This change _seems_ to reduce the frequency of the lock contention on TCP sockets in tests but doesn't fix it entirely and I'm not yet sure why. Leaving in Draft status while I debug further.

### CODE REVIEW

**Comment by DanGould** [comment]
Have you been able to reproduce the bug in a local environment or are you still relying on CI to produce it?

**Comment by spacebear21** [comment]
I'm able to reproduce locally but not as frequently as it occurs in CI (possibly just because CI runs 3x as many jobs).

**Comment by spacebear21** [comment]
Suggestion from @nothingmuch : Make the ohttp relay and payjoin directory port parameters optional, and simply find a free port during initialization if one is not provided.

**Comment by spacebear21** [comment]
Summarizing the problem:

Each v2 test initializes a OHTTP relay and payjoin directory. These processes run on a random "free" port selected by `find_free_port`. There is a race condition between finding a free port and actually initializing the relay/directory, resulting in the occasional "Address in use" error when the relay/directory process tries to bind to that port.

I tried the following approaches, none of which fix the issue:
- Use a Mutex to keep track of which ports have already been reserved. This doesn't work, afaict because it doesn't account for ports used by other processes (e.g. the redis and http clients).
- Spin up the test infrastructure (ohttp relay + directory) once and share the connections across tests. This doesn't work because tokio tests don't share a runtime (see https://github.com/payjoin/rust-payjoin/pull/388#issuecomment-2492030487)
- Don't specify a port and let the directory/ohttp relay select a free port on initialization ([as suggested](https://github.com/payjoin/rust-payjoin/pull/388#issuecomment-2489660918) by @nothingmuch ). The issue with this is that the tests need to know the relay & directory ports so that they know where to direct requests.

Other things I considered:
- Use named unix sockets instead of tcp but this seems quite involved...
- Use `testcontainers` to containerize the relay and directory processes, but this would necessitate making custom docker images and I'm not even sure it would work.

I've spent what feels like too many hours on this and am truly stumped... Happy to take another stab if anyone has new suggestions, but until then I'm putting this on the back burner.

**Comment by DanGould** [comment]
It can be painful to be stuck on something for a long time. Clearly what you've done so far is valuable and you've put lots of thought into the approach and potential remedies.

I was wondering most why @nothingmuch's comment didn't work and took [a stab at it myself](https://github.com/payjoin/rust-payjoin/commit/fe9d14668b68665aeb88bf18c17fcd8db9b58a6e)

> Don't specify a port and let the directory/ohttp relay select a free port on initialization (https://github.com/payjoin/rust-payjoin/pull/388#issuecomment-2489660918 by @nothingmuch ). The issue with this is that the tests need to know the relay & directory ports so that they know where to direct requests.

And the problem with this is that it's easy to get a deadlock passing both a handle and a port up from payjoin-directory. I'm sure there's a way to make this work, but my screwing around with it did not readily figure it out. The GET request in the stuck in the wait_for_service_ready loop seemed never to be able to `send()`. Easy to understand how you get into a black hole with this one.

In the meantime, might you share any scripts you used to test the tests? So we can pick this up at a time of rest?

@0xBEEFCAF3  just shared today how stoked he was on the clean cut-through interface. That took a long while to get right. This isn't the first time we've run into serious snags and I know we'll get through it.

**Comment by spacebear21** [comment]
> In the meantime, might you share any scripts you used to test the tests?

It pretty much amounts to this bad boy:

```zsh
export untilfail ()
{
  count=0
  while "$@"; do
    (( count++ ))
    echo "###### RUN COUNT: $count ######"
  done && say "failed after $count runs"
}
```

I run it like this and wait until the error occurs:

`untilfail cargo test --test integration --features=v2,danger-local-https -- --nocapture`

Note: on my old machine this fails reliably within ~40 runs - on my new/much faster machine it ~never fails.

**Comment by DanGould** [comment]
This must be a priority for test fixtures downstream in payjoin-ffi by binding to #422 so that we can replicate tests in bound implementations

**Comment by DanGould** [comment]
This is in a relatively clean state addressing flakiness, however, it can't be merged until https://github.com/payjoin/ohttp-relay/pull/41 is and a new `ohttp_relay-0.0.9` version is released. This version depends on the pull request's branch that introduces the same solution as we do on the directory to have it give US a port.

I'm not sure the status of #421 but I'm inclined to merge PRs in piecemeal since our major effort going forward is paying back tech debt. Best if we can target one issue and land it rather than block things that are huge packages of changes.

**Comment by DanGould** [comment]
I realize now when I requested your review I only pushed my changes to my own branch and not the one this PR is based off of. DOH! I still would have left the print statements in without your comment however, so that catch was good. We definitely DO want tracing.

I take it as a very good sign that [4cffe9c](https://github.com/payjoin/rust-payjoin/pull/388/commits/4cffe9cc31b3c9573f5240be4dd52ba2d1758176) passed all tests without flaking :D

**Comment by DanGould** [comment]
If it passes the smell test, I suppose we could merge with the git ref and update to ohttp-relay 0.0.9 once that is released without issue. It's not actually a hard block

File: payjoin/Cargo.toml [hunk 1]
```diff
@@ -41,7 +41,7 @@ serde_json = "1.0.108"
 bitcoind = { version = "0.36.0", features = ["0_21_2"] }
 http = "1"
 payjoin-directory = { path = "../payjoin-directory", features = ["_danger-local-https"] }
-ohttp-relay = "0.0.8"
+ohttp-relay = { version = "0.0.9", features = ["_test-util"] }
```
[the remark this answers is not in the document]
Comment (DanGould) [reply payjoin/Cargo.toml:44]: Specifying this in [`[dev-dependencies]`](https://doc.rust-lang.org/rust-by-example/testing/dev_dependencies.html) does exactly the behavior your suggest. It enables this feature only when testing

**Pull request merged**